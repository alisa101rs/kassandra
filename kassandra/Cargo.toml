[package]
name = "kassandra"
version.workspace = true
authors.workspace = true
description = "Single node kassandra implementation"
repository.workspace = true
license.workspace = true
readme.workspace = true
edition.workspace = true
keywords.workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
tokio = { version = "1", features = ["macros"], default-features = false }
tokio-util = { version = "0.7.8", features = ["codec"] }
eyre = "0.6.8"
futures = "0.3.28"
futures-util = { version = "0.3.28", features = ["sink"] }
nom = "7.1.3"
bytes = { version = "1.4.0", features = ["serde"] }
bitflags = "2"
indexmap = { version = "2.1.0", features = ["serde"] }
seahash = "4.1.0"
thiserror = "1.0.40"
ulid = "1.0.0"
uuid = { version = "1.10.0", features = ["serde"] }
serde = { version = "1", features = ["derive"] }
serde_bytes = "0.11.11"
serde_json = "1.0.107"
derive_more = "1.0.0"
chrono = { version = "0.4.27", features = ["serde"] }
bigdecimal = { version = "0.4.0", features = ["serde"] }
num-bigint = { version = "0.4.3", features = ["serde"] }
num_enum = { version = "0.6" }
ron = "0.8.0"
strum = { version = "0.25", features = ["derive"] }
integer-encoding = "4.0.0"

tracing = "0.1"
tracing-subscriber = { version = "0.3.17", features = ["env-filter"] }

[dev-dependencies]
insta = { version = "1.34.0" }
