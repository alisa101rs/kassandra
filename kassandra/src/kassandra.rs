use std::collections::HashMap;

use bytes::Bytes;
use eyre::Result;
use serde::Serialize;
use tracing::{instrument, Level};

use crate::{
    cql::query::QueryString,
    error::DbError,
    frame::{
        request::{
            batch::{Batch, BatchStatement},
            execute::Execute,
            query::Query,
        },
        response::{
            error::Error,
            event::{SchemaChangeEvent, SchemaChangeType},
            result::{Prepared, QueryResult, SchemaChange, SetKeyspace},
        },
    },
    snapshot::DataSnapshots,
    storage::{
        keyspace::{Keyspace, Strategy},
        system::{system_keyspace, system_schema_keyspace},
    },
};

#[derive(Debug, Clone, Serialize)]
pub struct Kassandra {
    use_keyspace: Option<String>,
    keyspace: HashMap<String, Keyspace>,
    #[serde(skip)]
    prepared: HashMap<u128, QueryString>,
}

impl Default for Kassandra {
    fn default() -> Self {
        Self::new()
    }
}

impl Kassandra {
    pub fn new() -> Self {
        let keyspace = [system_keyspace(), system_schema_keyspace()]
            .into_iter()
            .collect();

        Self {
            use_keyspace: None,
            prepared: HashMap::new(),
            keyspace,
        }
    }

    pub fn load_state<'a>(data: impl IntoIterator<Item = &'a [u8]>) -> Result<Self> {
        let mut kassandra = Self::new();

        for piece in data {
            let k = ron::de::from_bytes::<HashMap<String, Keyspace>>(piece)?;
            kassandra.keyspace.extend(k.into_iter());
        }

        Ok(kassandra)
    }

    #[instrument(level = Level::TRACE, skip(self), err, ret)]
    pub fn process(&mut self, query: Query<'_>) -> Result<QueryResult, Error> {
        match &query.query {
            QueryString::Use { keyspace, .. } => {
                self.use_keyspace = Some(keyspace.clone());

                Ok(QueryResult::SetKeyspace(SetKeyspace {
                    keyspace_name: keyspace.to_owned(),
                }))
            }
            QueryString::Select { keyspace, .. } => {
                let Some(keyspace) = keyspace.as_deref().or(self.use_keyspace.as_deref()) else {
                    Err(Error::new(DbError::Invalid, "keyspace must be specified"))?
                };
                let Some(keyspace) = self.keyspace.get_mut(keyspace) else {
                    Err(Error::new(DbError::Invalid, "keyspace does not exist"))?
                };
                Ok(keyspace.select(query)?)
            }
            QueryString::Insert { keyspace, .. } => {
                let Some(keyspace) = keyspace.as_deref().or(self.use_keyspace.as_deref()) else {
                    Err(Error::new(DbError::Invalid, "keyspace must be specified"))?
                };
                let Some(keyspace) = self.keyspace.get_mut(keyspace) else {
                    Err(Error::new(DbError::Invalid, "keyspace does not exist"))?
                };
                Ok(keyspace.insert(query)?)
            }
            QueryString::Delete { keyspace, .. } => {
                let Some(keyspace) = keyspace.as_deref().or(self.use_keyspace.as_deref()) else {
                    Err(Error::new(DbError::Invalid, "keyspace must be specified"))?
                };
                let Some(keyspace) = self.keyspace.get_mut(keyspace) else {
                    Err(Error::new(DbError::Invalid, "keyspace does not exist"))?
                };
                Ok(keyspace.delete(query)?)
            }

            QueryString::CreateKeyspace { .. } => Ok(self.create_keyspace(query)?),
            QueryString::CreateTable { keyspace, .. } => {
                let Some(keyspace) = keyspace.as_deref().or(self.use_keyspace.as_deref()) else {
                    Err(DbError::Invalid)?
                };
                let Some(keyspace) = self.keyspace.get_mut(keyspace) else {
                    Err(DbError::Invalid)?
                };

                Ok(keyspace.create_table(query)?)
            }
            QueryString::CreateType { .. } => {
                unimplemented!()
            }
        }
    }

    #[instrument(level = Level::TRACE, skip(self), err, ret)]
    pub fn execute(&mut self, execute: Execute<'_>) -> Result<QueryResult, Error> {
        let Execute {
            id,
            consistency,
            flags,
            data,
            page_size,
            paging_state,
            serial_consistency,
            default_timestamp,
        } = execute;
        let parsed_id = u128::from_be_bytes(id.try_into().unwrap());
        let Some(query) = self.prepared.get(&parsed_id).cloned() else {
            return Err(Error::new(
                DbError::Unprepared {
                    statement_id: Bytes::copy_from_slice(id)
                },
                "unprepared query"
            ))
        };

        let query = Query {
            query,
            raw_query: "",
            consistency,
            flags,
            data,
            page_size,
            paging_state,
            serial_consistency,
            default_timestamp,
        };

        self.process(query)
    }

    #[instrument(level = Level::TRACE, skip(self), err, ret)]
    pub fn process_batch(&mut self, batch: Batch<'_>) -> Result<QueryResult, Error> {
        let Batch {
            batch_type: _,
            consistency: _,
            serial_consistency: _,
            timestamp: _,
            values,
        } = batch;

        for query in values {
            match query {
                BatchStatement::Query(query) => {
                    let QueryResult::Void = self.process(query)? else {
                        unimplemented!("should be only void")
                    };
                }
                BatchStatement::Prepared(execute) => {
                    let QueryResult::Void = self.execute(execute)? else {
                        unimplemented!("should be only void")
                    };
                }
            }
        }

        Ok(QueryResult::Void)
    }

    fn create_keyspace(&mut self, query: Query<'_>) -> Result<QueryResult, DbError> {
        let QueryString::CreateKeyspace { keyspace,ignore_existence,  ..} = query.query else {
            unreachable!()
        };
        if self.keyspace.contains_key(&keyspace) {
            return if ignore_existence {
                Ok(QueryResult::Void)
            } else {
                Err(DbError::AlreadyExists {
                    keyspace,
                    table: "".to_string(),
                })
            };
        }

        self.keyspace.insert(
            keyspace.clone(),
            Keyspace {
                name: keyspace.clone(),
                strategy: Strategy::LocalStrategy,
                tables: Default::default(),
                user_defined_types: Default::default(),
            },
        );

        let event = SchemaChangeEvent::KeyspaceChange {
            change_type: SchemaChangeType::Created,
            keyspace_name: keyspace,
        };

        Ok(QueryResult::SchemaChange(SchemaChange { event }))
    }

    pub fn save_state(&self) -> Vec<u8> {
        ron::ser::to_string_pretty(&self.keyspace, Default::default())
            .unwrap()
            .into_bytes()
    }

    pub fn data_snapshot(&self) -> DataSnapshots {
        DataSnapshots::from_keyspaces(&self.keyspace)
    }

    #[instrument(level = Level::TRACE, skip(self), err, ret)]
    pub fn prepare(&mut self, query: QueryString) -> Result<QueryResult, Error> {
        self.prepare_with_id(query, ulid::Ulid::new().0)
    }

    pub fn prepare_with_id(&mut self, query: QueryString, id: u128) -> Result<QueryResult, Error> {
        if !matches!(
            query,
            QueryString::Insert { .. } | QueryString::Delete { .. } | QueryString::Select { .. }
        ) {
            return Err(Error::new(
                DbError::Invalid,
                "Only INSERT, DELETE or SELECT can be prepared",
            ));
        }
        let Some(keyspace) = query.keyspace().or(self.use_keyspace.as_deref()) else {
            Err(Error::new(DbError::Invalid, "keyspace must be specified"))?
        };
        let Some(keyspace) = self.keyspace.get_mut(keyspace) else {
            Err(Error::new(DbError::Invalid, "keyspace does not exist"))?
        };
        let (prepared_metadata, result_metadata) = keyspace.prepare(query.clone())?;

        self.prepared.insert(id, query);

        let prepared = Prepared {
            id,
            prepared_metadata,
            result_metadata,
        };

        Ok(QueryResult::Prepared(prepared))
    }

    pub fn use_keyspace(&mut self, ks: impl Into<String>) {
        self.use_keyspace = Some(ks.into());
    }
}
