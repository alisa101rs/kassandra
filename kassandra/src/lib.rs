pub mod cql;
pub mod error;
pub mod frame;
pub mod kassandra;
pub mod parse;
pub mod snapshot;
pub mod storage;
