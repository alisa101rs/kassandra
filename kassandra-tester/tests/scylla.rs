use std::collections::HashSet;

use insta::assert_yaml_snapshot;
use kassandra::kassandra::Kassandra;
use kassandra_tester::KassandraTester;
use scylla::{
    batch::{Batch, BatchType},
    FromRow, SessionBuilder, ValueList,
};

#[derive(Debug, ValueList, FromRow, Eq, PartialEq, Hash)]
struct TestData {
    key: String,
    c1: String,
    c2: String,
    value: String,
}

#[tokio::test]
async fn test_simple_test_data() -> eyre::Result<()> {
    let kassandra = Kassandra::new();

    let kassandra = KassandraTester::new(kassandra)
        .in_scope(|addr| async move {
            let s = SessionBuilder::new()
                .known_node(format!("{addr}"))
                .build()
                .await?;

            s
                .query("create keyspace if not exists test123 WITH REPLICATION = { 'class' : 'SimpleStrategy', 'replication_factor' : 1 }", ())
                .await.unwrap();

            s.query(
                "create table if not exists test123.test (key text, c1 text, c2 text, value text, PRIMARY KEY ((key), c1, c2))",
                (),
            )
                .await
                .unwrap();



            let values = [
                TestData {
                    key: "key".to_string(),
                    c1: "c1".to_string(),
                    c2: "c2".to_string(),
                    value: "value".to_string(),
                },
                TestData {
                    key: "key".to_string(),
                    c1: "c1".to_string(),
                    c2: "c22".to_string(),
                    value: "value".to_string(),
                },
                TestData {
                    key: "key".to_string(),
                    c1: "c12".to_string(),
                    c2: "c2".to_string(),
                    value: "value".to_string(),
                },
                TestData {
                    key: "key2".to_string(),
                    c1: "c1".to_string(),
                    c2: "c2".to_string(),
                    value: "value".to_string(),
                },
            ];

            for v in &values {
                s.query(
                    "insert into test123.test (key, c1, c2, value) values(?, ?, ?, ?)",
                    v,
                )
                    .await
                    .unwrap();
            }

            let r = s
                .query("select key, c1, c2, value from test123.test", &())
                .await
                .unwrap()
                .rows_typed::<TestData>()
                .unwrap();

            let rows = r.into_iter().collect::<Result<HashSet<_>, _>>()
                .unwrap();

            assert!(
                values.iter().all(|it| rows.contains(it))
            );

            let r = s
                .query(
                    "select key, c1, c2, value from test123.test where key=?",
                    &("key",),
                )
                .await
                .unwrap()
                .rows_typed::<TestData>()
                .unwrap();

            for r in r {
                let r = r.unwrap();
                assert_eq!(&r.key, "key")
            }

            let r = s
                .query(
                    "select key, c1, c2, value from test123.test where key=? AND c1 = ?",
                    &("key", "c1"),
                )
                .await
                .unwrap()
                .rows_typed::<TestData>()
                .unwrap();

            for r in r {
                let r = r.unwrap();
                assert_eq!(&r.key, "key");
                assert_eq!(&r.c1, "c1");
            }

            let r = s
                .query(
                    "select key, c1, c2, value from test123.test where key=? AND c1 = ? AND c2 = ?",
                    &("key", "c1", "c22"),
                )
                .await
                .unwrap()
                .maybe_first_row_typed::<TestData>()
                .unwrap();

            let r = r.unwrap();

            assert_eq!(&r.key, "key");
            assert_eq!(&r.c2, "c22");

            Ok::<(), eyre::Report>(())
        }).await?;

    assert_yaml_snapshot!(kassandra.data_snapshot());

    Ok(())
}

#[derive(Debug, ValueList, FromRow, Eq, PartialEq, Hash)]
struct TestBatchData {
    key: String,
    value: String,
}
#[tokio::test]
async fn test_simple_batch_data() -> eyre::Result<()> {
    let kassandra = Kassandra::new();

    let test = |addr| async move {
        let s = SessionBuilder::new()
            .known_node(format!("{addr}"))
            .build()
            .await?;

        s
            .query("create keyspace if not exists test WITH REPLICATION = { 'class' : 'SimpleStrategy', 'replication_factor' : 1 }", ())
            .await.unwrap();

        s.query(
            "create table if not exists test.t1 (key text, value text, PRIMARY KEY ((key)))",
            (),
        )
        .await
        .unwrap();

        s.query(
            "create table if not exists test.t2 (key text, value text, PRIMARY KEY ((key)))",
            (),
        )
        .await
        .unwrap();

        let mut batch = Batch::new(BatchType::Unlogged);
        batch.append_statement("insert into test.t1 (key, value) values(?, ?)");
        batch.append_statement("insert into test.t2 (key, value) values(?, ?)");
        let values = (
            TestBatchData {
                key: "key".to_string(),
                value: "value".to_string(),
            },
            TestBatchData {
                key: "key".to_string(),
                value: "value".to_string(),
            },
        );
        s.batch(&batch, values).await.unwrap();

        Ok::<_, eyre::Report>(())
    };
    let kassandra = KassandraTester::new(kassandra).in_scope(test).await?;

    assert_yaml_snapshot!(kassandra.data_snapshot());

    Ok(())
}

#[tokio::test]
async fn test_prepared() -> eyre::Result<()> {
    let kassandra = Kassandra::new();

    let test = |addr| async move {
        let s = SessionBuilder::new()
            .known_node(format!("{addr}"))
            .build()
            .await?;

        s
            .query("create keyspace if not exists test WITH REPLICATION = { 'class' : 'SimpleStrategy', 'replication_factor' : 1 }", ())
            .await.unwrap();

        s.query(
            "create table if not exists test.t1 (key text, value text, PRIMARY KEY ((key)))",
            (),
        )
        .await
        .unwrap();

        let insert = s
            .prepare("insert into test.t1 (key, value) values(?, ?)")
            .await
            .unwrap();
        let select = s
            .prepare("select key, value from test.t1 where key=?")
            .await
            .unwrap();

        let t = TestBatchData {
            key: "key".to_string(),
            value: "value".to_string(),
        };

        s.execute(&insert, t).await.unwrap();

        let v = s
            .execute(&select, ("key",))
            .await
            .unwrap()
            .maybe_first_row_typed::<TestBatchData>()
            .unwrap()
            .unwrap();
        assert_eq!(v.value, "value");

        Ok::<_, eyre::Report>(())
    };

    let kassandra = KassandraTester::new(kassandra).in_scope(test).await?;

    assert_yaml_snapshot!(kassandra.data_snapshot());

    Ok(())
}
