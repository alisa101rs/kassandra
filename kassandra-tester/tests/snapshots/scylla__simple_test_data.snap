---
source: kassandra-tester/tests/scylla.rs
expression: kassandra.data_snapshot()
---
test123:
  tables:
    test:
      rows:
        - partition_key: key
          clustering_key:
            - c1
            - c2
          data:
            c1: c1
            c2: c2
            key: key
            value: value
        - partition_key: key
          clustering_key:
            - c1
            - c22
          data:
            c1: c1
            c2: c22
            key: key
            value: value
        - partition_key: key
          clustering_key:
            - c12
            - c2
          data:
            c1: c12
            c2: c2
            key: key
            value: value
        - partition_key: key2
          clustering_key:
            - c1
            - c2
          data:
            c1: c1
            c2: c2
            key: key2
            value: value
